gobblin-core/src/main/java/gobblin/writer/http/AbstractAsyncDataWriter.java

```diff
@@ -0,0 +1,179 @@
+package gobblin.writer.http;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+import gobblin.writer.AsyncDataWriter;
+import gobblin.writer.WriteCallback;
+import gobblin.writer.WriteResponse;
+
+
+@ThreadSafe
+public abstract class AbstractAsyncDataWriter<D> implements AsyncDataWriter<D> {
+  public static int DEFAULT_BUFFER_CAPACITY = 10000;
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractAsyncDataWriter.class);
+
+  /** Main lock guarding all access */
+  private final ReentrantLock lock;
+
+  /** Condition for waiting takes */
+  private final Condition notEmpty;
+
+  /** Condition for waiting empty */
+  private final Condition isEmpty;
+
+  /** Condition for waiting puts */
+  private final Condition notFull;
+
+  /** Queue to batch records */
+  private final Queue<BufferedRecord<D>> buffer;
+
+  /** Queue capacity */
+  private final int capacity;
+
+  private final ExecutorService service;
+  private Future<?> recordProcessorFuture;
+
+  private volatile boolean isClosing;
+
+  public AbstractAsyncDataWriter(int capacity) {
+    service = Executors.newFixedThreadPool(1);
+    recordProcessorFuture = service.submit(new RecordProcessor());
+
+    lock = new ReentrantLock();
+    notEmpty = lock.newCondition();
+    isEmpty = lock.newCondition();
+    notFull =  lock.newCondition();
+    buffer = new ArrayDeque<>(capacity);
```

Comment:
The reasons are:
1. We want to move the logic of waiting for non empty queue to the `AbstractAsyncDataWriter` layer, for it is shared among all async data writers. The existing blocking queue doesn't provide this functionality without actually taking out an element.
2. Using a blocking queue, every `RequestBuilder` has to maintain some concurrent exception such as `Interrupted Exception`, which introduces too much common overhead and should be taken care of at a higher level as `AbstractAsyncDataWriter`
3. By maintaining a lock, we can introduce an `isEmpty` condition on which a `flush` call can wait. It's a simpler primitive than `CountUpDown` latch